### DISCUSSION TITLE
Determine Which Matrix Profile is Closest to a Target Time Series (Without Consensus Motif Search)

### DISCUSSION THREAD

**Discussion by jBavanati** [TDAmeritrade/stumpy#1047] [Help: Coding & Implementations] [answered]
Hi !
I perform AB Join operations on pairs of time series such as [1D,15D], [3D,15D], and [5D,15D]. This produces motifs for each pair, which is great so far.

Now, I want to determine which of the smaller-interval series (1D, 3D, 5D) has motifs most closely resembling the patterns in the 15D data.

What I’ve Tried:

* I considered consensus motif search but realized it’s not suitable because I’m not looking for the _same_ pattern across all time series. Instead, I’m focused on similarity to the 15D motifs alone.

Feature Request:

* Is there a built-in or recommended way to rank or compare the motifs from different matrix profiles (e.g., [1D,15D], [3D,15D], [5D,15D]) in terms of their closeness to the 15D time series?

Additional Notes:
* I am looking for a method or API functionality that could directly measure similarity between the discovered motifs and the target time series.
* Any advice or guidance would be greatly appreciated!

**Comment by seanlaw** [comment]
@jBavanati Thank you for your question and welcome to the STUMPY community.

> Is there a built-in or recommended way to rank or compare the motifs from different matrix profiles (e.g., [1D,15D], [3D,15D], [5D,15D]) in terms of their closeness to the 15D time series?

I don't know the answer to your question, but I wonder if you are looking for the `mpdist` function? See [this mpdist paper](https://www.cs.ucr.edu/~eamonn/MPdist_Expanded.pdf) and the corresponding [stumpy.mpdist](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.mpdist) function.

> I am looking for a method or API functionality that could directly measure similarity between the discovered motifs and the target time series.

As I'm sure you've already considered, the top (single) motif discovered in each matrix profile also comes with a distance and that would  be your "similarity" measure for which matrix profile has a subsequence that looks most similar to another subsequence in the target time series. However, it sounds like what you are interested in is what if we wanted to compare, say, the top 30 motifs found in each matrix profile that are computed against the same target time series. Is that correct? In that case, I believe that `mpdist` will get you there (or at least in the vicinity) by returning a pseudo time series-to-time-series distance. It's not an exact science but it's probably a good place to start. 

If I've misunderstood your question, please feel free to elaborate or correct me.

**Reply by jBavanati** [reply]
Thanks for the quick response !
I just realized I've overcomplicated the issue. What I'm looking for is a method to measure the degree of similarity between the two time series: the top motif found in 1D (or 3D or 5D) vector and the whole 15D time series (not the motifs found there). I think`mpdist` is an overkill for this. Is there a faster/ more lightweight way?

**Reply by JaKasb** [reply, marked as answer]
> the top motif found in 1D (or 3D or 5D) vector and the whole 15D time series   

`mass( query=top_motif_in_1D, timeseries = 15D_timeseries)`

**Reply by seanlaw** [reply]
Also see this `mass` tutorial: https://stumpy.readthedocs.io/en/latest/Tutorial_Pattern_Matching.html

**Comment by JaKasb** [comment]
Assuming 1D ... 15D are long time series instead of short motifs


```
for motif_i in list_of_motifs_from_15D :
    distance_profile_of_motif_i = mass( motif_i, 1D )
    np.min(distance_profile_of_motif_i)
```
https://stumpy.readthedocs.io/en/latest/api.html#stumpy.mass

You still need to build a python dictionary for all the motifs and timeseries, containing the minimum distance of the distance profiles. Then you can sort the dictionary by distance.

There is another way to compute the desired similarities, but it requires the full Distance-Matrix instead of the Matrix-Profile.
Usually the full distance matrix is too large to fit in RAM.

Hint: You can concatenate multiple timeseries into a single array, if you interleave a np.NaN value. 
By concatenating the indivdual timeseries into a single array, you can search your full data-corpus in a single API call.
The interleaved NaN value removes the concatenation-artefacts at the borders of the timesieries.
